Turn off warnings on Windows CI.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24331 Turn off warnings on Windows CI.**

Currently our logs are something like 40M a pop.  Turning off warnings and turning on verbose makefiles (to see the compile commands) reduces this to more like 8M. We could probably reduce log size more but verbose makefile is really useful and we'll keep it turned on for Windows.

Some findings:

1. Setting `CMAKE_VERBOSE_MAKEFILE` inside CMakelists.txt itself as suggested in https://github.com/ninja-build/ninja/issues/900#issuecomment-417917630 does not work on Windows. Setting `-DCMAKE_VERBOSE_MAKEFILE=1` does work (and we respect this environment variable.)
2. The high (`/W3`) warning level is by default on MSVC is due to cmake inserting this in the default flags. On recent versions of cmake, CMP0092 can be used to disable this flag in the default set. The string replace trick sort of works, but the standard snippet you'll find on the internet won't disable the flag from nvcc. I inspected the CUDA cmake code and verified it does respect CMP0092
3. `EHsc` is also in the default flags; this one cannot be suppressed via a policy. The string replace trick seems to work...
4. ... however, it seems nvcc implicitly inserts an `/EHs` after `-Xcompiler` specified flags, which means that if we add `/EHa` to our set of flags, you'll get a warning from nvcc. So we probably have to figure out how to exclude EHa from the nvcc flags set (EHs does seem to work fine.)
5. To suppress warnings in nvcc, you must BOTH pass `-w` and `-Xcompiler /w`. Individually these are not enough.

The patch applies these things; it also fixes a bug where nvcc verbose command printing doesn't work with `-GNinja`.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17131746](https://our.internmc.facebook.com/intern/diff/D17131746)

ezyang committed:
Turn off warnings on Windows CI.

Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@pytorchbot retest this please

ezyang commented:
@pytorchbot retest this please

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
more caffe2 wibbling on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Turn on verbose on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

xuhdev commented on CMakeLists.txt:
```diff
@@ -201,6 +203,10 @@ if (MSVC)
       string(REGEX REPLACE "/INCREMENTAL" "/INCREMENTAL:NO" ${flag_var} "${${flag_var}}")
     endif()
+    # Turn off warnings on Windows
+    string(REPLACE "/W3" "/w" ${flag_var} "${${flag_var}}")
```
Would this help? It looks like the CI is using the release build.

ezyang replied:
That would explain why this didn't work :)

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings
should help; we never read them anyway.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
Some notes about my findings:

1. Setting `CMAKE_VERBOSE_MAKEFILE` inside CMakelists.txt itself as suggested in https://github.com/ninja-build/ninja/issues/900#issuecomment-417917630 does not work on Windows. Setting `-DCMAKE_VERBOSE_MAKEFILE=1` does work (and we respect this environment variable.)
2. The high (`/W3`) warning level is by default on MSVC is due to cmake inserting this in the default flags. On recent versions of cmake, CMP0092 can be used to disable this flag in the default set. The string replace trick sort of works, but the standard snippet you'll find on the internet won't disable the flag from nvcc. I inspected the CUDA cmake code and verified it does respect CMP0092
3. `EHsc` is also in the default flags; this one cannot be suppressed via a policy. The string replace trick seems to work...
4. ... however, it seems nvcc implicitly inserts an `/EHs` after `-Xcompiler` specified flags, which means that if we add `/EHa` to our set of flags, you'll get a warning from nvcc. So we probably have to figure out how to exclude EHa from the nvcc flags set (EHs does seem to work fine.)
5. To suppress warnings in nvcc, you must BOTH pass `-w` and `-Xcompiler /w`. Individually these are not enough.

orionr reviewed (approved):
Looks reasonable!

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings and turning on verbose makefiles (to see the compile commands) reduces this to more like 8M. We could probably reduce log size more but verbose makefile is really useful and we'll keep it turned on for Windows.

1. Setting `CMAKE_VERBOSE_MAKEFILE` inside CMakelists.txt itself as suggested in https://github.com/ninja-build/ninja/issues/900#issuecomment-417917630 does not work on Windows. Setting `-DCMAKE_VERBOSE_MAKEFILE=1` does work (and we respect this environment variable.)
2. The high (`/W3`) warning level is by default on MSVC is due to cmake inserting this in the default flags. On recent versions of cmake, CMP0092 can be used to disable this flag in the default set. The string replace trick sort of works, but the standard snippet you'll find on the internet won't disable the flag from nvcc. I inspected the CUDA cmake code and verified it does respect CMP0092
3. `EHsc` is also in the default flags; this one cannot be suppressed via a policy. The string replace trick seems to work...
4. ... however, it seems nvcc implicitly inserts an `/EHs` after `-Xcompiler` specified flags, which means that if we add `/EHa` to our set of flags, you'll get a warning from nvcc. So we probably have to figure out how to exclude EHa from the nvcc flags set (EHs does seem to work fine.)
5. To suppress warnings in nvcc, you must BOTH pass `-w` and `-Xcompiler /w`. Individually these are not enough.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented on .jenkins/pytorch/win-test-helpers/build_pytorch.bat:
```diff
@@ -5,6 +5,7 @@ if "%DEBUG%" == "1" (
 )
 
 set PATH=C:\Program Files\CMake\bin;C:\Program Files\7-Zip;C:\ProgramData\chocolatey\bin;C:\Program Files\Git\cmd;C:\Program Files\Amazon\AWSCLI;%PATH%
+set CMAKE_VERBOSE_MAKEFILE=1
```
They're actually totally separate codepaths; `build_pytorch.bat` doesn't call `build_windows.bat`, and vice versa.

ezyang committed:
Update on "Turn off warnings on Windows CI."


Currently our logs are something like 40M a pop.  Turning off warnings and turning on verbose makefiles (to see the compile commands) reduces this to more like 8M. We could probably reduce log size more but verbose makefile is really useful and we'll keep it turned on for Windows.

1. Setting `CMAKE_VERBOSE_MAKEFILE` inside CMakelists.txt itself as suggested in https://github.com/ninja-build/ninja/issues/900#issuecomment-417917630 does not work on Windows. Setting `-DCMAKE_VERBOSE_MAKEFILE=1` does work (and we respect this environment variable.)
2. The high (`/W3`) warning level is by default on MSVC is due to cmake inserting this in the default flags. On recent versions of cmake, CMP0092 can be used to disable this flag in the default set. The string replace trick sort of works, but the standard snippet you'll find on the internet won't disable the flag from nvcc. I inspected the CUDA cmake code and verified it does respect CMP0092
3. `EHsc` is also in the default flags; this one cannot be suppressed via a policy. The string replace trick seems to work...
4. ... however, it seems nvcc implicitly inserts an `/EHs` after `-Xcompiler` specified flags, which means that if we add `/EHa` to our set of flags, you'll get a warning from nvcc. So we probably have to figure out how to exclude EHa from the nvcc flags set (EHs does seem to work fine.)
5. To suppress warnings in nvcc, you must BOTH pass `-w` and `-Xcompiler /w`. Individually these are not enough.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
